<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<title>Procedure</title>
<!-- 2019-05-23 Thu 10:59 -->
<meta  http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta  name="generator" content="Org-mode" />
<meta  name="author" content="Yttria" />
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center; }
  .todo   { font-family: monospace; color: red; }
  .done   { color: green; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  pre.src-sh:before    { content: 'sh'; }
  pre.src-bash:before  { content: 'sh'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-R:before     { content: 'R'; }
  pre.src-perl:before  { content: 'Perl'; }
  pre.src-java:before  { content: 'Java'; }
  pre.src-sql:before   { content: 'SQL'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.right  { text-align: center;  }
  th.left   { text-align: center;   }
  th.center { text-align: center; }
  td.right  { text-align: right;  }
  td.left   { text-align: left;   }
  td.center { text-align: center; }
  dt { font-weight: bold; }
  .footpara:nth-child(2) { display: inline; }
  .footpara { display: block; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  /*]]>*/-->
</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2013 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Procedure</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#sec-1">1. Make input directory (if necessary)</a></li>
<li><a href="#sec-2">2. PICRUSt2 Pipeline</a></li>
<li><a href="#sec-3">3. Converting to QIIME2 format</a></li>
</ul>
</div>
</div>
<p>
This file contains the procedure used to run PICRUSt2 metagenome predictions for the Global Coral Microbiome Project data.
</p>

<div id="outline-container-sec-1" class="outline-2">
<h2 id="sec-1"><span class="section-number-2">1</span> Make input directory (if necessary)</h2>
<div class="outline-text-2" id="text-1">
<div class="org-src-container">

<pre class="src src-sh"> # copy input files
# mdkir ../inpput
# cp ../../1_canonical_starting_files/all.biom ../input/all.biom
# cp ../../1_canonical_starting_files/all.seqs.fa ../input/all.seqs.fa
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-2" class="outline-2">
<h2 id="sec-2"><span class="section-number-2">2</span> PICRUSt2 Pipeline</h2>
<div class="outline-text-2" id="text-2">
<p>
The first step in the PICRUSt2 pipeline is to place sequences into a reference phylogeny by MSA + tree placement.
I had trouble with the <code>place_seqs.py</code> wrapper and instead called each
step manually.
</p>

<p>
This is the only really disgusting part.
</p>
<div class="org-src-container">

<pre class="src src-bash"># copied straight from the PICRUSt2 verbose output.
# BE CAREFUL!! This needs to be adjusted for your own conda install
hmmalign --trim --dna --mapali ~/miniconda/envs/picrust2/lib/python3.6/site-packages/picrust2/default_files/prokaryotic/pro_ref/pro_ref.fna.gz --informat FASTA -o ../working/query_align.stockholm ~/miniconda/envs/picrust2/lib/python3.6/site-packages/picrust2/default_files/prokaryotic/pro_ref/pro_ref.hmm ../input/all.seqs.fa

#(the above will work correctly if you just call place_seqs.py, but the next # step I found it useful to call directly
# note: --ref-msa arg comes from output of above
epa-ng --tree ~/miniconda/envs/picrust2/lib/python3.6/site-packages/picrust2/default_files/prokaryotics/pro_ref/pro_ref.tre                         \
--ref-msa ../working/ref_seqs_hmmalign.fasta \
--query output/intermediate/place_seqs/study_seqs_hmmalign.fasta    \
--chunk-size 2000 -T 8                                              \
-m ~/miniconda/envs/picrust2/lib/python3.6/site-packages/picrust2/default_files/prokaryotic/pro_ref/pro_ref.model                                   \
-w ../working/epa_out --filter-acc-lwr 0.5 --filter-max 20
# EPA-NG elapsed time (8cpu, 32gb EC2 instance): 22949s
</pre>
</div>

<p>
Then, the .jplace output from EPA-NG needs to be converted.
</p>
<div class="org-src-container">

<pre class="src src-bash">gappa analyze graft --jplace-path ../working/epa_out/epa_result.jplace --fully-resolve --out-dir ../working/epa_out/epa_result.newick
# picrust does this in two steps that I replicate here...
mv ../working/epa_out/epa_result.newick ../working/out.tre
</pre>
</div>

<p>
The next steps (hidden state prediction, metagenome prediction) aren't
so bad.
</p>

<div class="org-src-container">

<pre class="src src-bash">hsp.py -i 16S -t ../working/out.tre -o ../output/16S_predicted -p 8 -n
hsp.py -i EC  -t ../working/out.tre -o ../output/EC_predicted -p 8 -n --chunk_size 150
# KO had to be done on Atmosphere:
hsp.py -i KO  -t ../working/out.tre -o ../output/KO_predicted -p 4 --chunk_size 150
</pre>
</div>

<p>
Metagenome prediction is also straightforward although I think the
wiki argument list might have been slightly wrong.
</p>
<div class="org-src-container">

<pre class="src src-bash">metagenome_pipeline.py -i ../output/16S_predicted -o ../output/picrust_EC_metagenome
</pre>
</div>
<p>
There are a bunch of stratification and NSTI options and things.
Unfortunately, I couldn't get this step to run on the larger KO
predictions, not even on Atmosphere. 
</p>
</div>
</div>

<div id="outline-container-sec-3" class="outline-2">
<h2 id="sec-3"><span class="section-number-2">3</span> Converting to QIIME2 format</h2>
<div class="outline-text-2" id="text-3">
<p>
Converting the final metagenome files to qza table format is a
two-step process:
</p>

<div class="org-src-container">

<pre class="src src-bash"># convert from tsv to HDF5 BIOM
biom convert -i ../output/picrust_EC_metagenome/pred_metagenome_unstrat.tsv \
-o ../output/picrust_EC_metagenome/pred_metagenome_unstrat.biom             \
--table-type "Function table" --to-hdf5

# then from BIOM to qiime table
qiime tools import \
--input-path ../output/picrust_EC_metagenome/pred_metagenome_unstrat.biom \
--output-path ../output/picrust_EC_metagenome/ec_metagenome.qza           \
--type "FeatureTable[Frequency]"
</pre>
</div>

<p>
Same process with KO predictions, if available.
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Yttria</p>
<p class="date">Created: 2019-05-23 Thu 10:59</p>
<p class="creator"><a href="http://www.gnu.org/software/emacs/">Emacs</a> 25.2.2 (<a href="http://orgmode.org">Org</a> mode 8.2.10)</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
