{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import shutil\n",
    "import urllib.request\n",
    "\n",
    "from qiime2 import Artifact\n",
    "from qiime2.plugins.dada2.methods import denoise_single"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_file(url, local_filepath):\n",
    "    \"\"\"Download a file from a remote url and save to a local filepath\n",
    "    \n",
    "    url - the web address of the file you want to download as a string\n",
    "    local_filepath - the local filepath to which the file will be saved\n",
    "    \"\"\"\n",
    "    print(f\"Downloading file: {url}\")\n",
    "    with urllib.request.urlopen(url) as response, open(local_filepath, 'wb') as out_file:\n",
    "        shutil.copyfileobj(response, out_file)\n",
    "    \n",
    "    print(f\"Saved to local filepath: {local_filepath}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_directory(path):\n",
    "    \"\"\"Make a directory, but proceed without errors if it fails\n",
    "    path -- the path to the directory (e.g. \"../output/taxonomy_references\")\n",
    "    \"\"\"\n",
    "    try:\n",
    "        os.mkdir(path)\n",
    "    except OSError:\n",
    "        print (f\"Creation of directory {path} failed\")\n",
    "    else:\n",
    "        print (f\"Created the directory {path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_dir=os.path.abspath('../')\n",
    "mock_dir=working_dir+'/output/mock_communities'\n",
    "metadata_url_base='https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mocks=[12, 13, 14, 15, 16, 18, 19, 20, 21, 22]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating mock directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_12\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_12\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-12/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-12/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-12/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_12/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-t2x7o6kr/4281ff37-4095-4f94-8a14-d1c1596a853c/data /tmp/tmpw24up988/output.tsv.biom /tmp/tmpw24up988/track.tsv /tmp/tmpw24up988 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 12 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_13\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_13\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-13/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-13/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-13/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_13/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-pwobob9q/d20768b8-d22a-4c10-bd41-de0647664fc9/data /tmp/tmpvunyc0kc/output.tsv.biom /tmp/tmpvunyc0kc/track.tsv /tmp/tmpvunyc0kc 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 13 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_14\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_14\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-14/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-14/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-14/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_14/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-w8a7z4ot/53b34395-206c-4b83-9f8e-a8d6d4fea0a4/data /tmp/tmp_oakdoy7/output.tsv.biom /tmp/tmp_oakdoy7/track.tsv /tmp/tmp_oakdoy7 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 14 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_15\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_15\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-15/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-15/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-15/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_15/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-1pl8hxft/b257e001-31a6-4999-b882-77f13b6b9af2/data /tmp/tmp9gdpm5e4/output.tsv.biom /tmp/tmp9gdpm5e4/track.tsv /tmp/tmp9gdpm5e4 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 15 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_16\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_16\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-16/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-16/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-16/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_16/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-4bik5erp/15710978-a1b0-4e59-b544-f52c34beac02/data /tmp/tmpz_lrzhwo/output.tsv.biom /tmp/tmpz_lrzhwo/track.tsv /tmp/tmpz_lrzhwo 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 16 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_18\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_18\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-18/dataset-metadata.tsv\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-18/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-18/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_18/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-hxwaaeay/3c76b997-7115-4434-a5fc-3ab45c36b3c6/data /tmp/tmpzu3tg4_y/output.tsv.biom /tmp/tmpzu3tg4_y/track.tsv /tmp/tmpzu3tg4_y 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 18 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_19\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_19\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-19/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-19/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-19/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_19/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-zs4hzuvt/0ecafcb6-5b33-4e35-b026-43330724881a/data /tmp/tmplpt348pa/output.tsv.biom /tmp/tmplpt348pa/track.tsv /tmp/tmplpt348pa 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 19 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_20\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_20\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-20/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-20/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-20/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_20/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-xtlaj6vj/b12fe172-64d8-4b5c-bafa-60d9e7b0ad8a/data /tmp/tmpnmwn725e/output.tsv.biom /tmp/tmpnmwn725e/track.tsv /tmp/tmpnmwn725e 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 20 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_21\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_21\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-21/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-21/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-21/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_21/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-huktkz2q/3a0a599d-5cf4-4976-b68f-fe1b1f7b7671/data /tmp/tmppg8pdpyh/output.tsv.biom /tmp/tmppg8pdpyh/track.tsv /tmp/tmppg8pdpyh 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 21 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_22\n",
      "Created the directory /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_22\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-22/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-22/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-22/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/dsone/Documents/zaneveld/GCMP/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_22/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-s6q7v1rr/03956500-7053-42f6-87f4-b9d2a3de8741/data /tmp/tmplihdvva2/output.tsv.biom /tmp/tmplihdvva2/track.tsv /tmp/tmplihdvva2 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 22 imported\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#download the forward reads, import into qiime2, and create\n",
    "#sequence and feature table artifacts\n",
    "#this loop could take a while, depending on the power of your\n",
    "#cpu and your download speed\n",
    "if not os.path.exists(mock_dir):\n",
    "    print(f'Creating mock directory {mock_dir}')\n",
    "    make_directory(mock_dir)\n",
    "for mock in mocks:\n",
    "    mock_folder=mock_dir+'/mock_'+str(mock)\n",
    "    if not os.path.exists(mock_folder):\n",
    "        print(f'Creating mock folder {mock_folder}')\n",
    "        make_directory(mock_folder)\n",
    "    metadata_url=metadata_url_base+str(mock)+'/dataset-metadata.tsv'\n",
    "    metadata_filepath=mock_dir+'/metadata.tsv'\n",
    "    sample_metadata_url=metadata_url_base+str(mock)+'/sample-metadata.tsv'\n",
    "    sample_metadata_filepath=mock_dir+'/sample_metadata.tsv'\n",
    "    download_file(metadata_url, metadata_filepath)\n",
    "    download_file(sample_metadata_url, sample_metadata_filepath)\n",
    "    metadata=pd.read_csv(metadata_filepath, sep='\\t')\n",
    "    sample_metadata=pd.read_csv(sample_metadata_filepath, sep='\\t')\n",
    "    sample_name=sample_metadata['SampleID'][0]\n",
    "    forward_url=metadata['value'].values[2]\n",
    "    forward_filepath=mock_folder+'/1_1_L001_R1_001.fastq.gz'\n",
    "    #reverse_url=metadata['value'].values[3]\n",
    "    #reverse_filepath=mock_folder+'/'+str(mock)+'_reverse.fastq.gz'\n",
    "    #index_url=metadata['value'].values[4]\n",
    "    #index_filepath=mock_folder+'/'+str(mock)+'_index.'\n",
    "    download_file(forward_url, forward_filepath)\n",
    "    #download_file(reverse_url, reverse_filepath)\n",
    "    #importing a fastq file into qiime2 requires a manifest file and a metadata.yml file\n",
    "    if not os.path.exists(mock_folder+'/MANIFEST'):\n",
    "        with open(mock_folder+'/MANIFEST', 'w') as file:\n",
    "            file.write('sample-id,filename,direction\\n'+'sample1,'+forward_filepath+',forward')\n",
    "    if not os.path.exists(mock_folder+'/metadata.yml'):\n",
    "        with open(mock_folder+'/metadata.yml', 'w') as file:\n",
    "            pass\n",
    "    #now we need to import them into qiime\n",
    "    seqs = Artifact.import_data('SampleData[SequencesWithQuality]', '../output/mock_communities/mock_'+str(mock))\n",
    "    seqs.save(mock_folder+'/mock_'+str(mock)+'_sequences.qza')\n",
    "    table, sequences, stats=denoise_single(seqs, 0, n_threads=0)\n",
    "    table.save(mock_folder+'/mock_'+str(mock)+'_ft.qza')\n",
    "    sequences.save(mock_folder+'/mock_'+str(mock)+'_rep_seqs.qza')\n",
    "    print('mock', mock, 'imported\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
